### ISSUE TITLE
Possible Compilation or Run Time Issues with String-Based Attributes

### ISSUE THREAD

**Issue by ChrisDrozdowski** [steven-varga/h5cpp#22]
Hi,

I am finding compilation or run time issues when attempting to use `std::string` and `std::vector<std::string>` to read dataset attributes.

Can you confirm that you see these issues too?

Perhaps I am missing something in my code
 
```
#include <iostream>
#include <string>
#include <vector>
#include "h5cpp/all"

int main()
{
	// Create a file.
	h5::fd_t fd = h5::create("a.h5", H5F_ACC_TRUNC, h5::default_fcpl, h5::libver_bounds({ H5F_LIBVER_LATEST, H5F_LIBVER_LATEST }));

	// Add a simple dataset.
	std::vector<int> data = { 1,2,3,4,5 };
	h5::ds_t ds = h5::write(fd, "/test/path", data, h5::create_path | h5::utf8);

	// Employ the two methods for adding attributes to dataset to add std::string's.
	std::string str1 = "std::string #1";
	ds["attr1"] = str1;
	std::string str2 = "std::string #2";
	h5::awrite(ds, "attr2", str2);

	// Employ the two methods for added attributes to dataset to add std::vector<std::string>'s.
	std::vector<std::string> vs1 = { "Millicent", "Judith", "Jennifer" };
	ds["attr3"] = vs1;
	std::vector<std::string> vs2 = { "Duke", "Melvin", "Claude" };
	h5::awrite(ds, "attr4", vs2);

	// Read and dump the two std::string attributes.
	auto attr_1 = h5::aread<std::string>(ds, "attr1");
	std::cout << "attr1 : " << attr_1 << std::endl;

	auto attr_2 = h5::aread<std::string>(ds, "attr2");
	std::cout << "attr2 : " << attr_2 << std::endl;

	// Read and dump the two std::vector<std::string> attributes.
	auto attr_3 = h5::aread<std::vector<std::string>>(ds, "attr3");
	std::cout << "attr3 : ";
	for (auto v : attr_3)
		std::cout << v << ", ";
	std::cout << std::endl;

	auto attr_4 = h5::aread<std::vector<std::string>>(ds, "attr4");
	std::cout << "attr4 : ";
	for (auto v : attr_4)
		std::cout << v << ", ";
	std::cout << std::endl;
}
```

**Comment by ChrisDrozdowski** [comment]
Upon further investigation, it seems not possible to write `std::vector<std:string>` attributes.
One has to use `std::vector<const char*>` to write and `std::vector<char*>` to read. See code below.

Further, it appears that one can write string attributes with either `std::string` or `char*`.
But there seems no way to read either `std::string` or `char*` attributes. Compilations errors prevent even trying though this is commented on in H5Aread.hpp. See code below.
```
int main()
{
	h5::mute();

	// Create a file.
	h5::fd_t fd = h5::create("a.h5", H5F_ACC_TRUNC, h5::default_fcpl, h5::libver_bounds({ H5F_LIBVER_LATEST, H5F_LIBVER_LATEST }));

	// Add a simple dataset.
	std::vector<int> data = { 1,2,3,4,5 };
	h5::ds_t ds = h5::write(fd, "/test/path", data, h5::create_path | h5::utf8);

	// Employ methods for attributes to add std::string and char* attributes.
	// Verified- adds attributes to file.
	std::string str = "This is str_attr";
	ds["str_attr"] = str;
	const char* ch = "This is ch_attr";
	ds["ch_attr"] = ch;

	// Employ method for attributes to add vector<const char*> attribute.
	// Verified- adds attribute to file.
	std::vector<const char*> vch = { "Jane", "Judith", "June" };
	ds["vch_attr"] = vch;

	// Read and dump the std::string and char* attributes.
	// Both fail to compile at all. Expected because of comment for aread() in H5Aread.hpp.
	auto str_attr = h5::aread<std::string>(ds, "str_attr");
	std::cout << "str_attr : " << str_attr << std::endl;
	auto ch_attr = h5::aread<char*>(ds, "ch_attr");
	std::cout << "ch_attr : " << ch_attr << std::endl;

	// Read and dump the std::vector<char*> attribute.
	auto vch_attr = h5::aread<std::vector<char*>>(ds, "vch_attr");
	std::cout << "vch_attr : ";
	for (auto v : vch_attr)
		std::cout << v << ", ";
	std::cout << std::endl;
}
```

**Comment by steven-varga** [comment]
Good catch: no string read methods are provided. Some time ago I shifted the code base to a new compile time type system, and had been deliberating how to minimise overhead when it comes to string handling. As of now, re-evaluating the issue I don't see a fit for all solution, and it seems a `copy` operation from CAPI buffers is unavoidable.

For now I am re-adding the previously removed code block, and at some later point look into storing massive amount of strings with minimal overhead. 
Also this issue exists with general `h5::read<container<std::string>>()` operations, and will be solved with the same previously existing fetch + copy method.

**Comment by steven-varga** [comment]
Related to missing string support, fixed in commit cce9f0e795921c16fb193e55dae9dda38f1623af
